src/test/java/org/mqttbee/mqtt/codec/encoder/mqtt3/Mqtt3ConnectEncoderTest.java

```diff
@@ -17,113 +17,90 @@
 import org.mqttbee.mqtt.message.connect.MqttConnect;
 import org.mqttbee.mqtt.message.connect.MqttStatefulConnect;
 import org.mqttbee.mqtt.message.connect.mqtt3.Mqtt3ConnectView;
 import org.mqttbee.mqtt.message.publish.MqttWillPublish;
 
 import java.nio.ByteBuffer;
-import java.util.Objects;
 
+import static org.eclipse.paho.client.mqttv3.MqttConnectOptions.MQTT_VERSION_3_1_1;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.platform.commons.util.StringUtils.isNotBlank;
 
-class Mqtt3ConnectEncoderTest extends AbstractMqtt5EncoderTest {
-
-    private static final byte[] EXAMPLE_CONNECT = {
-            // FIXED HEADER
-            // packet type and flags
-            0x10,
-            // remaining length (16 Bytes)
-            0x10,
-            // VARIABLE HEADER
-            // protocol name
-            0x00, //msb
-            0x04, //lsb
-            'M', 'Q', 'T', 'T',
-            //protocol level
-            0x04,
-            //connect flags (only clean session is set)
-            0b0000_0010,
-            //keep alive (60)
-            0x00, 0x3c,
-            //clientId
-            0x00, 0x04, 'T', 'E', 'S', 'T'
-    };
+class Mqtt3ConnectEncoderTest extends AbstractMqtt3EncoderTest {
 
     Mqtt3ConnectEncoderTest() {
         super(code -> new Mqtt3ConnectEncoder(), false);
     }
 
-    @Test
-    void encode_SUCCESS() {
-        final MqttClientIdentifierImpl identifier = Objects.requireNonNull(MqttClientIdentifierImpl.from("TEST"));
-        final MqttConnect connect = Mqtt3ConnectView.delegate(60, true, null, null);
-        final MqttStatefulConnect connectWrapper = connect.createStateful(identifier, null);
-        encode(EXAMPLE_CONNECT, connectWrapper);
+    @CsvFileSource(resources = "/testParams/mqtt3/Connect.csv")
+    @ParameterizedTest(name = "Connect(\"{0}\", {1}, {2}, \"{3}\", \"{4}\", \"{5}\", \"{6}\")")
+    public void matchesPaho(
+            final String clientId,
+            final boolean cleanSession,
+            final int keepAliveInterval,
+            final String userName,
+            final String password,
+            final String willMessage,
+            final String willTopic,
+            final Integer willQos,
+            final Boolean willRetained
+    ) throws MqttException {
+        final boolean hasAuth = isNotBlank(userName) && isNotBlank(password);
+        final boolean hasWill = isNotBlank(willMessage)
+                && isNotBlank(willTopic)
+                && willQos != null
+                && willRetained != null;
+
+        final MqttSimpleAuth auth = !hasAuth ? null : new MqttSimpleAuth(
+                MqttUTF8StringImpl.from(userName),
+                ByteBuffer.wrap(password.getBytes(UTF8))
+        );
+        final MqttClientIdentifierImpl identifier = MqttClientIdentifierImpl.from(clientId);
+        final MqttWillPublish beeWill = !hasWill ? null : new MqttWillPublish(
```

Comment:
`Mqtt3PublishView.of(topic, payload, qos, retain).getWillDelegate();` would be easier